From cfbd831f0301b0bc3bfd2becace8e4d223bb5680 Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Mon, 16 May 2022 16:12:16 +0200
Subject: [PATCH] Integrate with Balena u-boot environment

Integrated u-boot with Balena environment
and added unzipping of the kernel image.
Increased default env size.

Upstream-Status: Inappropriate [configuration]
Signed-off-by: Vicentiu Galanopulo  <vicentiu@balena.io>
---
 configs/iot-gate-imx8.config    | 10 +++++++++-
 include/configs/iot-gate-imx8.h | 19 ++++++++++++++-----
 2 files changed, 23 insertions(+), 6 deletions(-)

diff --git a/configs/iot-gate-imx8.config b/configs/iot-gate-imx8.config
index ebaa232b88..cb4d5fe548 100644
--- a/configs/iot-gate-imx8.config
+++ b/configs/iot-gate-imx8.config
@@ -1,5 +1,5 @@
 CONFIG_TARGET_IOT_GATE_IMX8=y
-CONFIG_LOCALVERSION="-iot-gate-imx8-3.0"
+CONFIG_LOCALVERSION="-iot-gate-imx8-3.0 Balena build"
 CONFIG_DEFAULT_FDT_FILE="sb-iotgimx8-can.dtb"
 CONFIG_DEFAULT_DEVICE_TREE="iot-gate-imx8"
 CONFIG_DM_PCA953X=n
@@ -16,3 +16,11 @@ CONFIG_CMD_PCA953X=y
 CONFIG_DM_PCA953X=y
 CONFIG_SYS_PROMPT="IOT-GATE-iMX8 => "
 CONFIG_USB_ETHER_SMSC95XX=y
+CONFIG_CMD_UNZIP=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_SETEXPR=y
+CONFIG_PARTITION_UUIDS=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_FS_UUID=y
+
diff --git a/include/configs/iot-gate-imx8.h b/include/configs/iot-gate-imx8.h
index 9774949456..16cf6f7dcc 100644
--- a/include/configs/iot-gate-imx8.h
+++ b/include/configs/iot-gate-imx8.h
@@ -17,12 +17,15 @@
 #undef CONFIG_EXTRA_ENV_SETTINGS
 #undef CONFIG_BOOTCOMMAND
 
+#define CONFIG_ENV_SIZE 0x5000
+
 #define ENV_CONSOLE console=ttymxc2,115200 earlycon=ec_imx6q,0x30880000,115200
 #define CONFIG_EXTRA_ENV_SETTINGS		\
 	CONFIG_MFG_ENV_SETTINGS \
 	"autoload=off\0" \
 	"script=boot.scr\0" \
-	"image=Image\0" \
+	"image=Image.gz\0" \
+	"zip_addr=0x70480000\0" \
 	"fdt_addr=0x43000000\0"	\
 	"fdt_high=0xffffffffffffffff\0"	\
 	"initrd_high=0xffffffffffffffff\0" \
@@ -30,18 +33,24 @@
 	"bootargs_common="__stringify(ENV_CONSOLE)" net.ifnames=0\0" \
 	"emmc_ul=setenv boot_dev_str eMMC; run boot_try_msg; " \
 	"setenv iface mmc; setenv dev 2; setenv part 1;" \
-	"setenv bootargs ${bootargs_common} root=/dev/mmcblk2p2 " \
-	"${root_opt};\0" \
+	"setenv bootargs ${bootargs_common} " \
+	"${resin_kernel_root} ${root_opt} ${os_cmdline};\0" \
 	"usb_ul=setenv boot_dev_str USB; run boot_try_msg; usb reset; " \
 	"setenv iface usb; setenv dev 0; setenv part 1; " \
-	"setenv bootargs ${bootargs_common} root=/dev/sda2 ${root_opt};\0" \
+	"setenv bootargs ${bootargs_common} ${resin_kernel_root} ${root_opt} ${os_cmdline};\0" \
 	"ulbootscript=load ${iface} ${dev}:${part} ${loadaddr} ${script};\0" \
-	"ulimage=load ${iface} ${dev}:${part} ${loadaddr} ${image}\0" \
+	"ulimage=load ${iface} ${dev}:${part} ${zip_addr} ${image}; unzip ${zip_addr} ${loadaddr}\0" \
 	"ulfdt=load ${iface} ${dev}:${part} ${fdt_addr} ${fdt_file};\0" \
 	"bootscript=echo Running bootscript from ${boot_dev_str} ...; " \
 	"source;\0" \
 	"boot_try_msg=echo Trying to boot from ${boot_dev_str} ...;\0"
 #define CONFIG_BOOTCOMMAND \
+	"setenv resin_kernel_load_addr ${loadaddr};" \
+	"run resin_set_kernel_root; run set_os_cmdline; " \
+	"setenv usbdev ${resin_dev_index};" \
+	"setenv usbbootpart ${resin_boot_part};" \
+	"setenv mmcdev ${resin_dev_index};" \
+	"setenv mmcbootpart ${resin_boot_part};" \
 	"for src in usb_ul emmc_ul; do " \
 		"run ${src}; " \
 		"if run ulbootscript; then " \
-- 
2.17.1

